### ISSUE TITLE
[ENHANCEMENT] Harvester should display and honor the "virtual size" of uploaded qcow2 images

### ISSUE THREAD

**Issue by Martin-Weiss** [harvester/harvester#4905]
**Is your enhancement related to a problem? Please describe.**

In case we upload the SLE Micro 5.5 image (SLE-Micro.x86_64-5.5.0-Default-qcow-GM.qcow2) and create a new VM based on this image with default settings - the VM crashes during bootup.
Reason for this is that the default OS disk size in Harvester is 10 GB but the OS in this case requires 13 GB at least.
Because Harvester in this case "shrinks" below the minimum required - it causes the OS not to work.

` qemu-img info SLE-Micro.x86_64-5.5.0-Default-qcow-GM.qcow2`

```
image: SLE-Micro.x86_64-5.5.0-Default-qcow-GM.qcow2
file format: qcow2
virtual size: 20 GiB (21474836480 bytes)
disk size: 955 MiB
cluster_size: 65536
Format specific information:
    compat: 1.1
    compression type: zlib
    lazy refcounts: false
    refcount bits: 16
    corrupt: false
    extended l2: false

```

**Describe the solution you'd like**
Within a qcow2 image we can see the "virtual size" of an image - this should be taken as default for the OS image size at least. In case the admin selects something smaller - a warning should show up that there is a risk for problems if the OS disk is created smaller.

**Describe alternatives you've considered**
Manual adjustment of OS disk size during VM creation - this works but is not very user-friendly as Harvester does not give any idication of "more space required for the OS".

**Additional context**
Also the image overview in Harvester should show the "virtual size" in additon to the current "disk size" to help administrators.

**Comment by Vogtinator** [comment]
> Because Harvester in this case "shrinks" below the minimum required - it causes the OS not to work.

That's put very mildly. Harvester litererally deletes OS data from the disk in this case.

**Comment by Vicente-Cheng** [comment]
I thought that would be helpful and prevent the user from being confused about the VM crash.

WDYT?  cc @bk201, @markhillgit

**Comment by tserong** [comment]
I hit the same issue experimenting with the SLE Minimal VM image (_not_ the Cloud one):

```
> qemu-img info SLES15-SP5-Minimal-VM.x86_64-kvm-and-xen-QU1.qcow2
image: SLES15-SP5-Minimal-VM.x86_64-kvm-and-xen-QU1.qcow2
file format: qcow2
virtual size: 24 GiB (25769803776 bytes)
disk size: 255 MiB
cluster_size: 65536
Format specific information:
    compat: 1.1
    compression type: zlib
    lazy refcounts: false
    refcount bits: 16
    corrupt: false
    extended l2: false
Child node '/file':
    filename: SLES15-SP5-Minimal-VM.x86_64-kvm-and-xen-QU1.qcow2
    protocol type: file
    file length: 255 MiB (267845632 bytes)
    disk size: 255 MiB
```

If you don't know to set a larger volume size, the VM will just hang forever on boot waiting for the disk:

![sle-15-sp5-minimal-harvester-boot](https://github.com/harvester/harvester/assets/754700/67241c8c-2cdd-443a-a71d-0f88c389414e)

I think Martin's suggested solution seems reasonable (take the image's virtual size as the default if it's larger than harvester's default of 10GiB)

**Comment by tserong** [comment]
Same issue when using large ISO images (e.g. SLE-15-SP5-Full-x86_64-GM-Media1.iso). In this case, when trying to install SLE from an ISO that's been truncated to 10GiB, it manifests as this confusing error in the VM console:

![image](https://github.com/harvester/harvester/assets/754700/57030e8a-e681-45e6-9c17-e2be6d70428b)

**Comment by tserong** [comment]
I decided to see if I could figure out how to fix this.

For image storage, Harvester uses Longhorn's backing image feature (https://longhorn.io/docs/1.6.0/advanced-resources/backing-image/backing-image/). Any images uploaded into harvester are stored by Longhorn as backing images, which are separate to volumes. You can see them if you go "Setting: Backing Image" in the Longhorn UI.

When creating new VMs, Harvester puts some annotations in the VM metadata to indicate which volumes it wants, which storage classes to get them from, and how big the user wants those volumes to be, e.g.:

```
type: kubevirt.io.virtualmachine
metadata:
  namespace: default
  annotations:
    harvesterhci.io/volumeClaimTemplates: >-
    [
       {
          "metadata" : {
             "annotations" : {
                "harvesterhci.io/imageId" : "default/image-87dw7"
             },
             "name" : "sle-micro-55-disk-0-oz2bc"
          },
          "spec" : {
             "accessModes" : [
                "ReadWriteMany"
             ],
             "resources" : {
                "requests" : {
                   "storage" : "20Gi"
                }
             },
             "storageClassName" : "longhorn-image-87dw7",
             "volumeMode" : "Block"
          }
       },
       {
          "metadata" : {
             "annotations" : {
                "harvesterhci.io/imageId" : "default/image-vv6s6"
             },
             "name" : "sle-micro-55-disk-2-3p3r8"
          },
          "spec" : {
             "accessModes" : [
                "ReadWriteMany"
             ],
             "resources" : {
                "requests" : {
                   "storage" : "10Gi"
                }
             },
             "storageClassName" : "longhorn-image-vv6s6",
             "volumeMode" : "Block"
          }
       }
    ]
    network.harvesterhci.io/ips: '[]'
  labels:
    harvesterhci.io/creator: harvester
  name: sle-micro-55
[...]
```

Those storage classes in turn specify which backing images to use. Looking at the first storage class in the example above:

```
# kubectl -n longhorn-system get storageclass longhorn-image-87dw7 -o yaml
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  creationTimestamp: "2024-01-09T03:08:29Z"
  name: longhorn-image-87dw7
  resourceVersion: "51261"
  uid: 38e1657a-79fc-41e5-88e0-b51200d20d73
parameters:
  backingImage: default-image-87dw7
  migratable: "true"
  numberOfReplicas: "3"
  staleReplicaTimeout: "30"
provisioner: driver.longhorn.io
reclaimPolicy: Delete
volumeBindingMode: Immediate
```

Now let's look at the backing image:

```
# kubectl -n longhorn-system get lhbi default-image-87dw7 -o yaml
apiVersion: longhorn.io/v1beta2
kind: BackingImage
metadata:
  annotations:
    harvesterhci.io/imageId: default/image-87dw7
  creationTimestamp: "2024-01-09T03:08:29Z"
  finalizers:
  - longhorn.io
  generation: 12
  labels:
    longhorn.io/component: backing-image
    longhorn.io/managed-by: longhorn-manager
  name: default-image-87dw7
  namespace: longhorn-system
  resourceVersion: "48570880"
  uid: 1530d5cf-5dae-405e-8028-3a660609a91e
spec:
  checksum: ""
  disks:
    1deb74a7-9b57-4c80-ae2a-f4ac2214d78f: ""
    399b076f-2ab7-4138-8540-bab8103abde5: ""
    515a8af5-7664-4577-b65e-769890860769: ""
  sourceParameters:
    url: http://192.168.0.1/SLE-Micro.x86_64-5.5.0-Default-qcow-GM.qcow2
  sourceType: download
status:
  checksum: 1bbfc258229011fc4be34d35e823dab72431f577de5027d88f9236bbb8b786f25ca976422dfb55a02194215b0b717fbba033516c2b8c32b46c6587a580aaa748
  diskFileStatusMap:
    1deb74a7-9b57-4c80-ae2a-f4ac2214d78f:
      lastStateTransitionTime: "2024-01-09T07:36:04Z"
      message: ""
      progress: 100
      state: ready
    399b076f-2ab7-4138-8540-bab8103abde5:
      lastStateTransitionTime: "2024-02-13T08:51:23Z"
      message: ""
      progress: 100
      state: ready
    515a8af5-7664-4577-b65e-769890860769:
      lastStateTransitionTime: "2024-02-13T08:51:23Z"
      message: ""
      progress: 100
      state: ready
  diskLastRefAtMap: {}
  ownerID: harvester-node-1
  size: 1001652224
  uuid: def790f9
```

This is SLE-Micro.x86_64-5.5.0-Default-qcow-GM.qcow2. Note the size right at the end of the above chunk of YAML. 1001652224 / 1024 / 1024 == 955MiB, i.e. the size of the qcow2 image.

Note the _absence_ of any information about the virtual size of that qcow2 image, which is what we'd need to query to correctly populate the volume size in the Harvester UI before VM creation.

TL;DR: ISTM the first thing we need to do is make Longhorn expose the virtual size of qcow2 backing images...

**Comment by tserong** [comment]
Something like this should do the right thing for the Harvester VM creation page in the dashboard:

```diff
--- a/pkg/harvester/edit/kubevirt.io.virtualmachine/VirtualMachineVolume/type/vmImage.vue
+++ b/pkg/harvester/edit/kubevirt.io.virtualmachine/VirtualMachineVolume/type/vmImage.vue
@@ -140,6 +140,40 @@ export default {
 
     onImageChange() {
       const imageResource = this.$store.getters['harvester/all'](HCI.IMAGE)?.find( I => this.value.image === I.id);
+      const imageSize = imageResource.status?.size;
+
+      if (imageSize) {
+        // This is easy, but not great because for really small images (<1GiB,
+        // e.g. a 20MiB ignition image I was testing), the size comes out in the
+        // UI as 0.02GiB, but the input field doesn't ordinarily let you use
+        // decimals (it wants whole numbers).
+        this.$set(this.value, 'size', imageSize);
+
+        // One way of dealing with the decimal problem is to force the volume
+        // size to be the nearest whole number of GiB greater than or equal to
+        // the actual image size:
+        const imageSizeGiB = Math.ceil(imageSize / 1024 / 1024 / 1024);
+
+        this.$set(this.value, 'size', `${ imageSizeGiB }Gi`);
+
+        // Note that in both the above cases, the size is automatically reset
+        // any time a different image is selected, so if the user wants to
+        // override the value for any reason, they need to set the size _after_
+        // selecting an image, vs. the current behaviour where the size can be
+        // set before selecting an image, and choosing a different image doesn't
+        // change it.
+
+        // Once we get Longhorn backing images to expose the virtual size of
+        // qcow2 images (assuming that's done with a separate status value) we
+        // could maybe do something like this:
+        const imageVirtualSize = imageResource.status?.virtualSize;
+
+        if (imageVirtualSize) {
+          const imageVirtualSizeGiB = Math.ceil(imageVirtualSize / 1024 / 1024 / 1024);
+
+          this.$set(this.value, 'size', `${ Math.max(imageSizeGiB, imageVirtualSizeGiB) }Gi`);
+        }
+      }
 
       if (this.idx === 0) {
         if (/iso$/i.test(imageResource?.imageSuffix)) {
```

**Comment by tserong** [comment]
The logic for the Harvester VM creation page `onChange()` handler also needs to go into `getDiskRows()` in `pkg/harvester/mixins/harvester-vm/index.js`, so that correct defaults are set when you click "Create a Virtual Machine" on an image on the Images page. Something like the following. As a bonus, this also makes sure the bus and type are set correctly initially for ISO images (as is also done in the `onChange()` handler):

```diff
--- a/pkg/harvester/mixins/harvester-vm/index.js
+++ b/pkg/harvester/mixins/harvester-vm/index.js
@@ -400,15 +400,41 @@ export default {
       let out = [];
 
       if (_disks.length === 0) {
+        let bus = 'virtio';
+        let type = HARD_DISK;
+        let size = '10Gi';
+
+        const image = this.images.find( I => this.imageId === I.id);
+        const imageSize = image?.status?.size;
+
+        if (imageSize) {
+          const imageSizeGiB = Math.ceil(imageSize / 1024 / 1024 / 1024);
+
+          size = `${ imageSizeGiB }Gi`;
+
+          const imageVirtualSize = image.status?.virtualSize;
+
+          if (imageVirtualSize) {
+            const imageVirtualSizeGiB = Math.ceil(imageVirtualSize / 1024 / 1024 / 1024);
+
+            size = `${ Math.max(imageSizeGiB, imageVirtualSizeGiB) }Gi`;
+          }
+        }
+
+        if (/iso$/i.test(image?.imageSuffix)) {
+          bus = 'sata';
+          type = CD_ROM;
+        }
+
         out.push({
           id:               randomStr(5),
           source:           SOURCE_TYPE.IMAGE,
           name:             'disk-0',
           accessMode:       'ReadWriteMany',
-          bus:              'virtio',
+          bus,
           volumeName:       '',
-          size:             '10Gi',
-          type:             HARD_DISK,
+          size,
+          type,
           storageClassName: '',
           image:            this.imageId,
           volumeMode:       'Block',
```

Still need to figure out the Longhorn bit...

**Comment by tserong** [comment]
We also need something like this piece of glue in the main harvester repo:

```diff
--- a/deploy/charts/harvester-crd/templates/harvesterhci.io_virtualmachineimages.yaml
+++ b/deploy/charts/harvester-crd/templates/harvesterhci.io_virtualmachineimages.yaml
@@ -123,6 +123,9 @@ spec:
                 type: integer
               storageClassName:
                 type: string
+              virtualSize:
+                format: int64
+                type: integer
             type: object
         required:
         - spec

--- a/pkg/apis/harvesterhci.io/v1beta1/image.go
+++ b/pkg/apis/harvesterhci.io/v1beta1/image.go
@@ -77,6 +77,9 @@ type VirtualMachineImageStatus struct {
 	// +optional
 	Size int64 `json:"size,omitempty"`
 
+	// +optional
+	VirtualSize int64 `json:"virtualSize,omitempty"`
+
 	// +optional
 	StorageClassName string `json:"storageClassName,omitempty"`
 
--- a/pkg/controller/master/image/backing_image_controller.go
+++ b/pkg/controller/master/image/backing_image_controller.go
@@ -51,6 +51,8 @@ func (h *backingImageHandler) OnChanged(_ string, backingImage *lhv1beta2.Backin
 			harvesterv1beta1.ImageImported.Message(toUpdate, status.Message)
 			toUpdate.Status.Progress = status.Progress
 			toUpdate.Status.Size = backingImage.Status.Size
+			// Requires update to "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2".BackingImageStatus
+			toUpdate.Status.VirtualSize = backingImage.Status.VirtualSize
 		} else if status.Progress != toUpdate.Status.Progress {
 			harvesterv1beta1.ImageImported.Unknown(toUpdate)
 			harvesterv1beta1.ImageImported.Reason(toUpdate, "Importing")
```

**Comment by lanfon72** [comment]
Verified this enhancement has been applied.


Test Information
----
* Environment: **qemu/KVM 3 nodes**
* Harvester Version: **v1.4-1305f11-head**
* **ui-source** Option: **Auto**

### Verify Steps
1. Install Harvester with any nodes
2. Upload qcow2 images
3. Uploaded qcow2 images should display virtual size in the GUI which same as executing command `qemu-img info` in the command line.
